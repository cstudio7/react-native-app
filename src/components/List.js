import React from 'react';
import { Text, SectionList, StyleSheet } from 'react-native';
import createSection from '../modules/createSection/createSection';
import ListItem from '../containers/ListItem';
import Spacing from '../constants/Spacing';
import ListSectionHeader from './ListSectionHeader'

class List extends React.Component {
  renderSectionHeader = ({ section: { title } }) => <ListSectionHeader title={title} />;

  renderItem = ({ item }) => (
    <ListItem name={item} gender={this.props.gender} page={this.props.page} />
  );

  render() {
    const { names } = this.props;
    let sections = [];
    if (this.props.page === 'favorites') {
      sections = createSection(
        names.filter(nameObject => nameObject.isFavorite)
      );
    } else {
      sections = createSection(names);
    }

    return (
      <SectionList
        style={styles.list}
        renderSectionHeader={this.renderSectionHeader}
        renderItem={this.renderItem}
        sections={sections}
        stickySectionHeadersEnabled={true}
        keyExtractor={item => item.id}
      />
    );
  }
}

const styles = StyleSheet.create({
  list: {
    paddingLeft: Spacing.padding2,
    paddingRight: Spacing.padding2
  }
});

export default List;
