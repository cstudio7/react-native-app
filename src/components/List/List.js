import React from 'react';
import ListItemComponent from '../../containers/ListItem';
import ListEmptyComponent from '../ListEmptyComponent/ListEmptyComponent';
import { Spacing } from '../../constants';

import { Animated, FlatList, Platform, View, StyleSheet } from 'react-native';
import { Item, Input, Icon } from 'native-base';
import NotFound from '../NotFound';

const searchFormHeight = 72;

class List extends React.Component {
  scroll = new Animated.Value(0);
  headerY;

  constructor(props) {
    super(props);
    this.scrollEventHandler = props.scrollEvent;

    this.headerY = Animated.multiply(
      Animated.diffClamp(this.scroll, 0, searchFormHeight),
      -1
    );

    this.state = {
      target: null
    };

    this.clearSearchForm = this.clearSearchForm.bind(this);
    this.onChangeTextHandler = this.onChangeTextHandler.bind(this);
  }

  renderItem = ({ item }) => (
    <ListItemComponent
      navigation={this.props.navigation}
      name={item}
      gender={this.props.route.key}
      screen={this.props.screen}
    />
  );

  renderListEmptyComponent = () => (
    <ListEmptyComponent screen={this.props.screen} />
  );

  keyExtractor = (item, index) => item.name + index;

  onChangeTextHandler(text) {
    this.setState({ target: text });
  }

  clearSearchForm() {
    this.setState({
      target: null
    });
  }

  render() {
    let { data } = this.props;

    if (this.state.target) {
      data = data.filter(item =>
        item.name.toLowerCase().startsWith(this.state.target.toLowerCase())
      );
    }

    return (
      <View style={styles.container}>
        {this.props.screen !== 'Favorites' && (
          <Animated.View
            style={[
              styles.searchForm,
              {
                transform: [
                  {
                    translateY: this.headerY
                  }
                ]
              }
            ]}>
            <Item rounded>
              <Icon active name="md-search" />
              <Input
                placeholder="Поиск"
                value={this.state.target}
                onChangeText={this.onChangeTextHandler}
              />
              {this.state.target ? (
                <Icon active name="md-close" onPress={this.clearSearchForm} />
              ) : null}
            </Item>
          </Animated.View>
        )}

        <Animated.FlatList
          scrollEventThrottle={16}
          bounces={false}
          showsVerticalScrollIndicator={false}
          style={styles.animatedFlatList}
          onScroll={Animated.event(
            [{ nativeEvent: { contentOffset: { y: this.scroll } } }],
            { useNativeDriver: true }
          )}
          overScrollMode="never"
          contentContainerStyle={
            this.props.screen !== 'Favorites' &&
            styles.animatedFlatListContentContainer
          }
          renderItem={this.renderItem}
          data={data}
          keyExtractor={this.keyExtractor}
          ListEmptyComponent={
            this.state.target && !data.length ? (
              <NotFound />
            ) : (
              <ListEmptyComponent screen={this.props.screen} />
            )
          }
          removeClippedSubviews={Platform.OS === 'android'}
          onScrollBeginDrag={this.scrollEventHandler}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white'
  },
  animatedFlatList: {
    zIndex: 0,
    height: '100%',
    elevation: -1
  },
  animatedFlatListContentContainer: {
    paddingTop: searchFormHeight - Spacing.base
  },
  searchForm: {
    paddingTop: Spacing.base,
    paddingLeft: Spacing.base,
    paddingRight: Spacing.base,
    backgroundColor: 'white',
    paddingBottom: Spacing.base,
    backgroundColor: 'white',
    width: '100%',
    position: 'absolute',
    elevation: 0,
    flex: 1,
    zIndex: 1
  }
});

export default List;
