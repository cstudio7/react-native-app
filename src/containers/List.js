import { connect } from 'react-redux';
import List from '../components/List/List';
import { makeGetSections } from '../redux/namesReducer/selectors';

const makeMapStateToProps = () => {
  const getSections = makeGetSections();
  const mapStateToProps = (state, props) => ({
    sections: getSections(state, props)
  });
  return mapStateToProps;
};

export default connect(
  makeMapStateToProps,
  {}
)(List);
