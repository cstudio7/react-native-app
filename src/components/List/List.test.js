import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import List from './List';

const middlewares = [];
const mockStore = configureStore(middlewares);
const store = mockStore({});

it('renders list correctly', () => {
  const data = [
    { id: 1, name: 'artur' },
    { id: 2, name: 'alex' },
    { id: 3, name: 'pierre' }
  ];

  const props = {
    screen: 'Favorites',
    data,
    route: { key: 'male' },
    scrollEvent: jest.fn()
  };

  const tree = renderer
    .create(
      <Provider store={store}>
        <List {...props} />
      </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('shows blank state if list is empty', () => {
  const props = {
    data: [],
    route: { key: 'female' },
    scrollEvent: jest.fn()
  };

  const tree = renderer
    .create(
      <Provider store={store}>
        <List {...props} />
      </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
