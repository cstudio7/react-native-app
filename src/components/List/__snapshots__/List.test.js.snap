// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders list correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    bounces={false}
    contentContainerStyle={false}
    data={
      Array [
        Object {
          "id": 1,
          "name": "artur",
        },
        Object {
          "id": 2,
          "name": "alex",
        },
        Object {
          "id": 3,
          "name": "pierre",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    overScrollMode="never"
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={10}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "elevation": -1,
        "height": "100%",
        "zIndex": 0,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 0.5,
                "borderColor": "#c9c9c9",
                "flexDirection": "row",
                "marginLeft": 16,
                "paddingBottom": 13,
                "paddingRight": 16,
                "paddingTop": 13,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": null,
                  "alignSelf": null,
                  "flex": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "marginLeft": 10,
                      "marginRight": 10,
                    },
                    Object {
                      "marginLeft": 0,
                    },
                  ]
                }
                uppercase={false}
              >
                artur
              </Text>
              <Text
                note={true}
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "#808080",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "200",
                      "marginLeft": 10,
                      "marginRight": 10,
                    },
                    Object {
                      "marginLeft": 0,
                    },
                  ]
                }
                uppercase={false}
              />
            </View>
            <View
              style={
                Object {
                  "alignItems": "flex-end",
                  "alignSelf": "center",
                  "flex": 0.28,
                  "paddingBottom": null,
                  "paddingLeft": null,
                  "paddingRight": null,
                  "paddingTop": null,
                }
              }
            >
              <View
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 12,
                      },
                      Array [
                        Object {
                          "color": "#000",
                          "fontSize": 30,
                        },
                        Object {
                          "alignSelf": null,
                          "color": "#c9c8cd",
                          "fontSize": 22,
                        },
                        Object {
                          "color": "#673AB7",
                          "fontSize": 22,
                        },
                      ],
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 0.5,
                "borderColor": "#c9c9c9",
                "flexDirection": "row",
                "marginLeft": 16,
                "paddingBottom": 13,
                "paddingRight": 16,
                "paddingTop": 13,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": null,
                  "alignSelf": null,
                  "flex": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "marginLeft": 10,
                      "marginRight": 10,
                    },
                    Object {
                      "marginLeft": 0,
                    },
                  ]
                }
                uppercase={false}
              >
                alex
              </Text>
              <Text
                note={true}
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "#808080",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "200",
                      "marginLeft": 10,
                      "marginRight": 10,
                    },
                    Object {
                      "marginLeft": 0,
                    },
                  ]
                }
                uppercase={false}
              />
            </View>
            <View
              style={
                Object {
                  "alignItems": "flex-end",
                  "alignSelf": "center",
                  "flex": 0.28,
                  "paddingBottom": null,
                  "paddingLeft": null,
                  "paddingRight": null,
                  "paddingTop": null,
                }
              }
            >
              <View
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 12,
                      },
                      Array [
                        Object {
                          "color": "#000",
                          "fontSize": 30,
                        },
                        Object {
                          "alignSelf": null,
                          "color": "#c9c8cd",
                          "fontSize": 22,
                        },
                        Object {
                          "color": "#673AB7",
                          "fontSize": 22,
                        },
                      ],
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 0.5,
                "borderColor": "#c9c9c9",
                "flexDirection": "row",
                "marginLeft": 16,
                "paddingBottom": 13,
                "paddingRight": 16,
                "paddingTop": 13,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": null,
                  "alignSelf": null,
                  "flex": 1,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000",
                      "fontFamily": "System",
                      "fontSize": 16,
                      "marginLeft": 10,
                      "marginRight": 10,
                    },
                    Object {
                      "marginLeft": 0,
                    },
                  ]
                }
                uppercase={false}
              >
                pierre
              </Text>
              <Text
                note={true}
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "color": "#808080",
                      "fontFamily": "System",
                      "fontSize": 14,
                      "fontWeight": "200",
                      "marginLeft": 10,
                      "marginRight": 10,
                    },
                    Object {
                      "marginLeft": 0,
                    },
                  ]
                }
                uppercase={false}
              />
            </View>
            <View
              style={
                Object {
                  "alignItems": "flex-end",
                  "alignSelf": "center",
                  "flex": 0.28,
                  "paddingBottom": null,
                  "paddingLeft": null,
                  "paddingRight": null,
                  "paddingTop": null,
                }
              }
            >
              <View
                accessible={true}
                isTVSelectable={true}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "opacity": 1,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": undefined,
                        "fontSize": 12,
                      },
                      Array [
                        Object {
                          "color": "#000",
                          "fontSize": 30,
                        },
                        Object {
                          "alignSelf": null,
                          "color": "#c9c8cd",
                          "fontSize": 22,
                        },
                        Object {
                          "color": "#673AB7",
                          "fontSize": 22,
                        },
                      ],
                      Object {
                        "fontFamily": "Ionicons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                    ]
                  }
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`shows blank state if list is empty 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "elevation": 0,
        "flex": 1,
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
        "position": "absolute",
        "transform": Array [
          Object {
            "translateY": -0,
          },
        ],
        "width": "100%",
        "zIndex": 1,
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "transparent",
          "borderBottomWidth": 1,
          "borderColor": "#D9D5DC",
          "borderLeftWidth": 1,
          "borderRadius": 30,
          "borderRightWidth": 1,
          "borderTopWidth": 1,
          "borderWidth": 1,
          "flexDirection": "row",
          "marginLeft": 2,
          "opacity": 1,
        }
      }
    >
      <Text
        active={true}
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Array [
              Object {
                "color": "#000",
                "fontSize": 30,
              },
              Object {
                "color": "#000",
                "fontSize": 24,
                "paddingLeft": 10,
                "paddingRight": 8,
              },
            ],
            Object {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
      <TextInput
        allowFontScaling={true}
        editable={true}
        onChangeText={[Function]}
        placeholder="Поиск"
        placeholderTextColor="#575757"
        style={
          Object {
            "color": "#000",
            "flex": 1,
            "fontSize": 17,
            "height": 50,
            "paddingLeft": 8,
            "paddingRight": 5,
            "top": 1.5,
          }
        }
        underlineColorAndroid="rgba(0,0,0,0)"
        value={null}
      />
      <Text
        active={true}
        allowFontScaling={false}
        onPress={[Function]}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 12,
            },
            Array [
              Object {
                "color": "#000",
                "fontSize": 30,
              },
              Object {
                "color": "#000",
                "fontSize": 24,
                "paddingLeft": 10,
                "paddingRight": 8,
              },
            ],
            Object {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    bounces={false}
    contentContainerStyle={
      Object {
        "paddingTop": 64,
      }
    }
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    overScrollMode="never"
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={10}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "elevation": -1,
        "height": "100%",
        "zIndex": 0,
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "paddingTop": 50,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9e9e9e",
              "fontSize": 16,
              "fontWeight": "500",
            }
          }
        >
          Вы пока ничего не добавили в избранное
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
